src/v/storage/disk_log_impl.cc

```diff
@@ -785,21 +766,19 @@ ss::future<> disk_log_impl::gc(compaction_config cfg) {
           gclog.trace,
           "[{}] skipped log deletion, internal topic",
           config().ntp());
-        return ss::make_ready_future<>();
+        co_return std::nullopt;
     }
-    if (cfg.max_bytes) {
-        size_t max = cfg.max_bytes.value();
-        vlog(
-          gclog.debug,
-          "[{}] retention max bytes: {}, current partition size: {}",
-          config().ntp(),
-          max,
-          _probe.partition_size());
-        if (!_segs.empty() && _probe.partition_size() > max) {
-            return garbage_collect_max_partition_size(cfg);
-        }
+
+    auto max_offset_by_size = size_based_gc_max_offset(cfg);
+    auto max_offset_by_time = time_based_gc_max_offset(cfg);
+
+    auto max_offset = std::max(max_offset_by_size, max_offset_by_time);
+
+    if (max_offset) {
+        co_return co_await request_eviction_until_offset(*max_offset);
```

Comment:
This may be a behavioral change worth leaving out. At the very least release noting on this version.

Reply:
Sorry, but I don't follow. How does the behaviour change?

Reply:
Ah sorry for not being clear: this seems to pull in behavior introduced in https://github.com/redpanda-data/redpanda/pull/9487 where we apply both the size- and space-based retention at once. Perhaps we should uphold the older behavior where we check size-based offsets first, else time-based? Or consider backporting #9487 (though it might be a tough sell because it changes how retention works)

Reply:
The two retention policies weren't really applied sequentially as the inline segment removal code was dead.
Both retentions would notify the eviction monitor separately though, which could result in two truncations.

However the order of the two truncations was not guaranteed. For instance, size based could request truncation at offset `X` and time based could request truncation at `Y`. Truncation at `Y` could happen first, which makes truncation at `X` a no-op (this is very similar to the new behaviour; after this PR). The other way around is also possible, which results in two sequential truncations (I guess the expected behaviour you had in mind).

The point is that the previous behaviour was basically undefined and this change makes it defined. I think it's fine to backport this.

Reply:
Hm maybe this comment was misplaced, but I'm referring to the difference behavior introduced compared to here

https://github.com/redpanda-data/redpanda/pull/9611/commits/f3f71684475118b810352e9beb40ca3324bc354c#diff-8caa0c9719de25bd209fa674959d9327cd55237de2fb4683761acf9bcc6de217L780-L791

where we would previously only ever time-based-truncate if we could not size-based-truncate. Or is that also a non-issue because of some dead code paths?

Reply:
I understand your point now. I don't really want to change this backport to preserve that behaviour though.
The previous behaviour feels like a bug to me, but maybe I'm missing something.

Reply:
Yeah I'm okay with including it -- my point was just to make sure this isn't an accident and that the https://github.com/redpanda-data/redpanda/pull/9487 release note follows this PR as well.

Reply:
Matt, Noah and I had a chat and the conclusion was that we shouldn't backport the behaviour change from https://github.com/redpanda-data/redpanda/pull/9487. I updated the PR to pick up some commits from https://github.com/redpanda-data/redpanda/pull/9487 to make things apply nicely and I tweaked [4800111](https://github.com/redpanda-data/redpanda/pull/9611/commits/48001113f08e19c4e7ee361ce3cb68e3919bcd96) to preserve the old behaviour.

Reply:
SGTM. Thanks for resolving this!